File: roboplan_examples/python/example_ik.py
#!/usr/bin/env python3

import sys
import time
import tyro
import xacro

import numpy as np
import pinocchio as pin

from common import MODELS
from roboplan.core import Scene, JointConfiguration, CartesianConfiguration
from roboplan.example_models import get_package_share_dir
from roboplan.simple_ik import SimpleIkOptions, SimpleIk
from roboplan.viser_visualizer import ViserVisualizer


def main(
    model: str = "ur5",
    max_iters: int = 100,
    step_size: float = 0.25,
    max_linear_error_norm: float = 0.001,
    max_angular_error_norm: float = 0.001,
    check_collisions: bool = True,
    host: str = "localhost",
    port: str = "8000",
):
    """
    Run the IK example with the provided parameters.


    Parameters:
        model: The name of the model to use.
        max_iters: Maximum number of iterations for the IK solver.
        step_size: Integration step size for the IK solver.
        max_linear_error_norm: The maximum linear error norm for the IK solver.
        max_angular_error_norm: The maximum angular error norm for the IK solver.
        check_collisions: Whether to check for collisions when solving IK.
        host: The host for the ViserVisualizer.
        port: The port for the ViserVisualizer.
    """

    if model not in MODELS:
        print(f"Invalid model requested: {model}")
        sys.exit(1)

    model_data = MODELS[model]
    package_paths = [get_package_share_dir()]

    # Pre-process with xacro. This is not necessary for raw URDFs.
    urdf_xml = xacro.process_file(model_data.urdf_path).toxml()
    srdf_xml = xacro.process_file(model_data.srdf_path).toxml()

    # Specify argument names to distinguish overloaded Scene constructors from python.
    scene = Scene(
        "test_scene",
        urdf=urdf_xml,
        srdf=srdf_xml,
        package_paths=package_paths,
        yaml_config_path=model_data.yaml_config_path,
    )
    q_full = scene.getCurrentJointPositions()
    q_indices = scene.getJointGroupInfo(model_data.default_joint_group).q_indices

    # Create a redundant Pinocchio model just for visualization.
    # When Pinocchio 4.x releases nanobind bindings, we should be able to directly grab the model from the scene instead.
    model = pin.buildModelFromXML(urdf_xml)
    collision_model = pin.buildGeomFromUrdfString(
        model, urdf_xml, pin.GeometryType.COLLISION, package_dirs=package_paths
    )
    visual_model = pin.buildGeomFromUrdfString(
        model, urdf_xml, pin.GeometryType.VISUAL, package_dirs=package_paths
    )

    viz = ViserVisualizer(model, collision_model, visual_model)
    viz.initViewer(open=True, loadModel=True, host=host, port=port)

    # Set up an IK solver
    options = SimpleIkOptions(
        group_name=model_data.default_joint_group,
        max_iters=max_iters,
        step_size=step_size,
        max_linear_error_norm=max_linear_error_norm,
        max_angular_error_norm=max_angular_error_norm,
        check_collisions=check_collisions,
    )
    ik_solver = SimpleIk(scene, options)

    start = JointConfiguration()
    start.positions = np.array(model_data.starting_joint_config)[q_indices]

    goals = []
    transform_controls = []
    for ee_name in model_data.ee_names:
        goal = CartesianConfiguration()
        goal.base_frame = model_data.base_link
        goal.tip_frame = ee_name
        goals.append(goal)

    solution = JointConfiguration()

    # Create interactive markers.
    def solve_ik(_):
        for goal, controls in zip(goals, transform_controls):
            goal.tform = pin.SE3(
                pin.Quaternion(controls.wxyz[[1, 2, 3, 0]]), controls.position
            ).homogeneous
        result = ik_solver.solveIk(goals, start, solution)
        if result:
            q_full[q_indices] = solution.positions
            viz.display(q_full)
            scene.setJointPositions(q_full)
            start.positions = solution.positions

    for ee_name in model_data.ee_names:
        controls = viz.viewer.scene.add_transform_controls(
            f"/ik_markers/{ee_name}",
            depth_test=False,
            scale=0.2,
            disable_sliders=True,
            visible=True,
        )
        controls.on_update(solve_ik)
        transform_controls.append(controls)

    # Create a marker reset button.
    reset_button = viz.viewer.gui.add_button("Reset Marker")

    @reset_button.on_click
    def reset_position(_):
        for goal, controls in zip(goals, transform_controls):
            fk_tform = scene.forwardKinematics(
                scene.getCurrentJointPositions(), goal.tip_frame
            )
            controls.position = fk_tform[:3, 3]
            controls.wxyz = pin.Quaternion(fk_tform[:3, :3]).coeffs()[[3, 0, 1, 2]]

    random_button = viz.viewer.gui.add_button("Randomize Pose")

    @random_button.on_click
    def randomize_position(pos):
        q_full = scene.getCurrentJointPositions()
        q_rand = scene.randomCollisionFreePositions()[q_indices]
        q_full[q_indices] = q_rand
        scene.setJointPositions(q_full)
        viz.display(q_full)
        start.positions = q_rand
        reset_position(pos)

    # Display the arm and marker at the starting position, then sleep forever.
    randomize_position(None)
    try:
        while True:
            time.sleep(10.0)
    except KeyboardInterrupt:
        pass


if __name__ == "__main__":
    tyro.cli(main)

File: roboplan_examples/cpp/example_ik.cpp
#include <filesystem>
#include <memory>
#include <vector>

#include <roboplan/core/scene.hpp>
#include <roboplan_example_models/resources.hpp>
#include <roboplan_simple_ik/simple_ik.hpp>

using namespace roboplan;

int main(int /*argc*/, char* /*argv*/[]) {
  // Set up the scene
  const auto share_prefix = example_models::get_package_share_dir();
  const auto model_prefix = share_prefix / "roboplan_example_models" / "models";
  const auto urdf_path = model_prefix / "ur_robot_model" / "ur5_gripper.urdf";
  const auto srdf_path = model_prefix / "ur_robot_model" / "ur5_gripper.srdf";
  const std::vector<std::filesystem::path> package_paths = {share_prefix};

  auto scene = std::make_shared<Scene>("example_ik_scene", urdf_path, srdf_path, package_paths);

  // Set up and solve IK
  SimpleIkOptions options;
  options.group_name = "arm";
  options.step_size = 0.25;
  options.max_linear_error_norm = 0.001;
  options.max_angular_error_norm = 0.001;
  options.check_collisions = true;
  auto ik_solver = SimpleIk(scene, options);

  auto model = scene->getModel();
  auto q_tgt = pinocchio::neutral(model);
  q_tgt[0] += 0.1;
  q_tgt[2] -= 0.1;
  q_tgt[4] -= 0.25;

  const auto goal = CartesianConfiguration{
      .base_frame = "base",
      .tip_frame = "tool0",
      .tform = scene->forwardKinematics(q_tgt, "tool0"),
  };

  JointConfiguration start;
  start.positions = pinocchio::neutral(model);

  JointConfiguration solution;
  const auto success = ik_solver.solveIk(goal, start, solution);

  if (!success) {
    std::cout << "IK solve failed!\n";
    return 0;
  }

  std::cout << "IK solve succeeded\n  Result: " << solution.positions.transpose() << "\n";

  return 0;
}
